src/lsm/README.md

```diff
@@ -0,0 +1,111 @@
+
+* At the end of every beat, there is space in mutable table for the next beat.
+* The manifest info for the compaction's input tables is updated at the end of the half-bar.
+* The manifest info for the compaction's output tables is updated during the compaction.
+* The manifest is compacted at the end of every beat.
+* The compactions' output tables are not visible until the compaction has finished.
+
+1. (first) down beat of the bar:
+    * Assert no compactions are currently running.
+    * Allow level visible table counts to overflow if needed.
+    * Start even level compactions if there's any tables to compact.
+
+2. (second) up beat of the bar:
+    * Finish ticking running even-level compactions.
+    * Assert on callback completion that all compactions are complete.
+
+3. (third) down beat of the bar:
+    * Assert no compactions are currently running.
+    * Start odd level compactions if there are any tables to compact, and only if we must.
+    * Compact the immutable table if it contains any sorted values (it could be empty).
+
+4. (fourth) last beat of the bar:
+    * Finish ticking running odd-level and immutable table compactions.
+    * Assert on callback completion that all compactions are complete.
+    * Assert on callback completion that no level's visible table count overflows.
+    * Flush, clear, and sort mutable table values into immutable table for next bar.
+
+
+## Snapshots
+### Snapshots and Compaction
+
+Consider the half-bar compaction beginning at op=`X` (`12`), with `lsm_batch_multiple=M` (`8`).
+Each half-bar contains `N=M/2` (`4`) beats. The next half-bar begins at `Y=X+N` (`16`).
+
+During the compaction `X` (op=`X…Y-1`; `12…15`), each commit prefetches from the snapshot equal to
+the first op of the compaction. As shown, they continue to query the old (input) tables.
+
+During the compaction `X`:
+- `snapshot_max` of each input table is truncated to `Y-1` (`15`).
+- `snapshot_min` of each output table is initialized to `Y` (`16`).
+
+```
+  0   4   8  12  16  20  24  (op, snapshot)
+  ┼───┬───┼───┬───┼───┬───┼
+              ####
+  ····────────X────────····  (input  tables, before compaction)
+  ····────────────           (input  tables,  after compaction)
+                  Y────····  (output tables,  after compaction)
+```
+
+Beginning from the next op after the compaction (`Y`; `16`):
+- The output tables of the above compaction `X` are visible.
+- The input tables of the above compaction `X` are invisible.
+- Therefore, it will lookup from the output tables, but ignore the input tables.
+- Commits must not prefetch from the output tables of `X` before the compaction half-bar
+  has finished, since those tables are incomplete.
+
+At this point the input tables can be removed if they are invisible to all persistent snapshots.
+
+
+### Snapshot Values
+
+The on-disk tables visible to a snapshot `B` do not contain the updates from the commit with op `B`.
```

Comment:
Why not?

Reply:
Because we tried that before and it was very hard to reason about :wink:. Surprisingly (for me anyway) "which ops' updates are in this snapshot" isn't useful at this level of LSM-tree abstraction.

Reply:
So the mutable table is flushed before op `B` to produce snapshot `B`?

Reply:
No, the mutable table flushed before `B` would produce snapshot `C`. But now that I write that out, I realize it is confusing. I'm going to see whether it would make sense to shift things by a bar so they match, or if it would be confusing in a different way. :thinking: